### DISCUSSION TITLE
Raspotify no longer starts on my MagicMirror

### DISCUSSION THREAD

**Discussion by cacy82** [dtcooper/raspotify#530] [Q&A] [answered]
Hello everybody,
I made some MagicMirror updates and now raspotify no longer works as it should.

If I open Spotify from my smartphone or pc, the raspberry is seen among the devices. If I choose it, the album cover with all the data also appears on the MagicMirror, including the exact moment in which the song is located at that moment, but if I press play the cover disappears and passes the playback on the device from which I am starting it.

How can I solve?

EDIT: I Have Spotify Premium
![raspotify su MM](https://user-images.githubusercontent.com/96691751/160288629-88f3afbb-e837-4a47-b4bc-24c63a87bebe.JPG)
![spotify da pc](https://user-images.githubusercontent.com/96691751/160288630-67c84260-0059-4efb-9b7b-b6620a41ef84.JPG)

**Comment by JasonLG1979** [comment]
It's really hard to say? There's a lot of missing info.

## *What version of Raspotify/librespot are you running?*


To get that version info:
```
librespot -V
```

The current amd64 build outputs:
```
librespot 0.3.1 650d41b (Built on 2022-03-20, Build ID: yPu3748c, Profile: release)
```

The librespot version, commit (`librespot 0.3.1 650d41b`) and "Built on" date (`Built on 2022-03-20`) will match if you're running the most current version of Raspotify.

## *What Debian based distro are you using?*

To get some very basic system version info:
```
lsb_release -d
```

Raspotify is only compatible with Debian >= 11 (Bulleye) based distros. Debian 10 (Buster) based distros are incompatible and not supported at all.

## *What do the Raspotify service logs say?*

This will output the Raspotify service systemd logs for the current boot:
```
sudo journalctl -u raspotify -b
```

**Comment by cacy82** [comment]
Hi,

sorry for missing some information; I'll report them below

`librespot 0.3.1 650d41b (Built on 2022-03-20, Build ID: 8JMqPd04, Profile: release)`

`Raspbian GNU/Linux 11 (bullseye)`

`LOG`
-- Journal begins at Sat 2022-01-01 23:43:48 CET, ends at Mon 2022-03-28 08:11:25 CEST. --
mar 27 17:36:21 raspberrypi systemd[1]: Started Raspotify (Spotify Connect Client).
mar 27 17:36:24 raspberrypi librespot[528]: [2022-03-27T15:36:24Z ERROR librespot] Discovery Error: Setting up dns-sd failed: No such device (os error 19)
mar 27 17:36:24 raspberrypi systemd[1]: raspotify.service: Main process exited, code=exited, status=1/FAILURE
mar 27 17:36:24 raspberrypi systemd[1]: raspotify.service: Failed with result 'exit-code'.
mar 28 08:02:11 raspberrypi systemd[1]: raspotify.service: Scheduled restart job, restart counter is at 1.
mar 28 08:02:11 raspberrypi systemd[1]: Stopped Raspotify (Spotify Connect Client).
mar 28 08:02:11 raspberrypi systemd[1]: Started Raspotify (Spotify Connect Client).
mar 28 08:10:14 raspberrypi librespot[848]: [2022-03-28T06:10:14Z WARN  librespot_core::mercury] error 400 for uri hm://autoplay-enabled/query?uri=
mar 28 08:10:14 raspberrypi librespot[848]: [2022-03-28T06:10:14Z ERROR librespot_connect::spirc] AutoplayError: MercuryError
mar 28 08:10:17 raspberrypi librespot[848]: Failed to get home directory.
mar 28 08:10:17 raspberrypi librespot[848]: [2022-03-28T06:10:17Z ERROR librespot_playback::player] Audio Sink Error Invalid Parameters: <AlsaSink> Device hw:CARD=vc4hdmi,DEV=0 Unsupported Format S16LE (S16), ALSA function 'snd_pcm_hw_pa>
mar 28 08:10:18 raspberrypi systemd[1]: raspotify.service: Main process exited, code=exited, status=1/FAILURE
mar 28 08:10:18 raspberrypi systemd[1]: raspotify.service: Failed with result 'exit-code'.
mar 28 08:10:28 raspberrypi systemd[1]: raspotify.service: Scheduled restart job, restart counter is at 2.
mar 28 08:10:28 raspberrypi systemd[1]: Stopped Raspotify (Spotify Connect Client).
mar 28 08:10:28 raspberrypi systemd[1]: Started Raspotify (Spotify Connect Client).

**Comment by cacy82** [comment]
Anyone can help me?

**Comment by JasonLG1979** [comment]
```
mar 28 08:10:17 raspberrypi librespot[848]: [2022-03-28T06:10:17Z ERROR librespot_playback::player] Audio Sink Error Invalid Parameters: Device hw:CARD=vc4hdmi,DEV=0 Unsupported Format S16LE (S16), ALSA function 'snd_pcm_hw_pa>
```

 Device `hw:CARD=vc4hdmi,DEV=0` does not support `S16`, the default `format` used in `librespot`. You can see what formats that output supports with `librespot --device ?`.
 
The troubleshooting wiki is [Here](https://github.com/dtcooper/raspotify/wiki/Troubleshooting).

**Comment by cacy82** [comment]
I used `librespot --device ?` and see that my device support `S16 format`

![image](https://user-images.githubusercontent.com/96691751/161370285-51d5e0b0-5ff0-42c5-aba5-87f5ac07c651.png)

i'm going on `/ect/Raspotify/conf` with `nano` and i past here the screenshoot:

![image](https://user-images.githubusercontent.com/96691751/161371035-869152e9-bb8a-4bbb-8982-289e0a27e019.png)

I don't understand where i'm wrong :(

**Reply by JasonLG1979** [reply]
I'm not sure either? It says it supports `S32` try that. If that doesn't work try `S24` and `S24_3`?

I wrote that code in `librespot`, `--device ?` basically just asks the devices which formats they support. Apparently the `vc4hdmi` driver is lying because it says it supports `S16` then chokes on it?

**Reply by JasonLG1979** [reply]
What does the device on the other end of the HDMI cable support? That's also pretty important I'd imagine? I'm not sure off the top of my head? You might look into that.

**Comment by cacy82** [comment]
I don't know, it's a Samsung TV monitor,
After i've use the command `aplay -Dhw:2 --dump-hw-params /usr/share/sounds/alsa/Front_Right.wav` and I see this:

![image](https://user-images.githubusercontent.com/96691751/161372465-5a57fd0b-4bc4-435c-b2d5-73bb378d52af.png)

**Reply by JasonLG1979** [reply]
Make sure the TV is set to PCM sound.

**Comment by cacy82** [comment]
The audio on tv works without problems,
it recognizes the youtube module and the googleassistant module and the audio play.
Raspotify First it worked without problems, then since I did the updates which suggested Raspbian it stopped working,
but only raspotify, the other audio function works without problems.

If I try `speaker-test -c2 -l1` the sound is heard and i see this:

![image](https://user-images.githubusercontent.com/96691751/161372889-ba45cc02-718d-4e3e-ba90-87112b30cd71.png)

**Reply by JasonLG1979** [reply]
> it recognizes the youtube module and the googleassistant module and the audio play

If you're using several other services that play audio I would not advise that any one of them directly access the hardware. You're better off routing everything though dmix (ALSA's built in software mixer) the reason being that only one program at a time can directly access the hardware. With several programs trying to directly access the hardware, basically willy-nilly, you're bound to run into issues where 2 or more programs try to access the hardware at the same time which will cause audio crashes.

**Reply by JasonLG1979** [reply]
The [Troubleshooting wiki page](https://github.com/dtcooper/raspotify/wiki/Troubleshooting) has instructions on how to set a specific sound device as the default (which will route though dmix) under the *"My Raspberry Pi does not use my sound card/device!"* section. After you've done that just tell each service to use the default output. Raspotify defaults to the default sound device so all you have to do after that for Raspotify is just omit the device setting.

**Comment by cacy82** [comment]
I tried to disable the other services that use the audio, but the result has not changed, and I also tried to follow that section of the Troubleshooting (both with the other services active and disabled) but the result does not change.

I was thinking one thing, but it's not that it's all related to the first error that appears in the log:

`mar 27 17:36:24 raspberrypi librespot[528]: [2022-03-27T15:36:24Z ERROR librespot] Discovery Error: Setting up dns-sd failed: No such device (os error 19)`

I don't understand much about programming, so I rely on you!

**Reply by JasonLG1979** [reply]
> mar 27 17:36:24 raspberrypi librespot[528]: [2022-03-27T15:36:24Z ERROR librespot] Discovery Error: Setting up dns-sd failed: No such device (os error 19)

`os error 19` means something is write protected from my limited Googling of it? I'm not sure what that means? If you're already providing your Spotify username and password you can disable discovery. The only drawback is Raspotify will only be visible to that user. If you're the only one who plans on using it that's not really a problem though.

That doesn't have anything to do with this error though.

``` 
mar 28 08:10:17 raspberrypi librespot[848]: [2022-03-28T06:10:17Z ERROR librespot_playback::player] Audio Sink Error Invalid Parameters: Device hw:CARD=vc4hdmi,DEV=0 Unsupported Format S16LE (S16), ALSA function 'snd_pcm_hw_pa>
```

> I don't understand much about programming, so I rely on you!

I'm just really good at obsessively Googling things :smile:

**Comment by cacy82** [comment]
It wouldn't be a problem because the raspberry is mounted on the magic mirror I have in my bathroom, I don't think anyone else can have access to it. :D

I am trying them all, but they cannot be solved, yet before the updates required by the raspbian it worked wonders :(

I also tried deleting raspotify and reinstalling it, but the problem remains :(

**Comment by cacy82** [comment, marked as answer]
UPDATE!!!!
I managed to get it started!

I went into `/etc/raspotify/conf` and I changed the parameter `LIBRESPOT_DEVICE =` from `hdmi: CARD = vc4hdmi, DEV = 0` with `sysdefault: CARD = vc4hdmi` 

And now working perfectly!
